kaitaistruct.py

```diff
@@ -15,51 +20,171 @@
+            s._read()
+            return s
 
     @classmethod
-    def from_bytes(cls, buf):
-        return cls(KaitaiStream(BytesIO(buf)))
+    def from_file(cls, file: typing.Union[Path, str], use_mmap:bool=True):
+        return cls.from_any(file, use_mmap=use_mmap)
 
-    @classmethod
-    def from_io(cls, io):
-        return cls(KaitaiStream(io))
 
+class IKaitaiDownStream(ABC):
+    __slots__ =("_io",)
 
-class KaitaiStream(object):
-    def __init__(self, io):
-        self._io = io
-        self.align_to_byte()
+    def __init__(self, _io: typing.Any):
+        self._io = _io
+
+    @abstractmethod
+    @property
+    def is_entered(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def __enter__(self):
+        raise NotImplementedError()
+
+    def __exit__(self, *args, **kwargs):
+        if self.is_entered():
+            self._io.__exit__(*args, **kwargs)
+            self._io = None
+
+    def __del__(self):
+        self.__exit__(None, None)
+
+
+class KaitaiFileSyscallDownStream(IKaitaiDownStream):
+    __slots__ = ()
+    def __init__(self, path: typing.Union[Path, str]):
+        super().__init__(Path(path))
+
+    @property
+    def is_entered(self):
+        return isinstance(self._io, _IOBase)
 
     def __enter__(self):
+        self._io = open(self.path).__enter__()
+        return self
+
+
+class KaitaiFileMapDownStream(IKaitaiFileDownStream):
+    __slots__ = ("file",)
+    def __init__(self, path: Path):
+        super().__init__(None)
+        self.file = KaitaiFileSyscallDownStream(path)
+
+    @property
+    def is_entered(self):
+        return isinstance(self._io, mmap.mmap)
+
+    def __enter__(self):
+        self.file = self.file.__enter__()
+        self._io = mmap.mmap(self.file.file.fileno(), 0, access=mmap.ACCESS_READ).__enter__()
         return self
 
     def __exit__(self, *args, **kwargs):
-        self.close()
+        super().__exit__(*args, **kwargs)
+        if self.file is not None:
+            self.file.__exit__(*args, **kwargs)
+            self.file = None
+
+
+def getFileDownStream(path: Path, *args, use_mmap: bool=True, **kwargs) -> IKaitaiFileDownStream:
+    if use_mmap:
+        cls = KaitaiFileMapDownStream
+    else:
+        cls = KaitaiFileSyscallDownStream
+    return cls(path, *args, **kwargs)
+
+
+class KaitaiBytesDownStream(IKaitaiBytesDownStream):
+    __slots__ = ()
+    def __init__(self, data: bytes):
+        super().__init__(data)
+
+    @property
+    def is_entered(self):
+        return isinstance(self._io, BytesIO)
+
+    def __enter__(self):
+        self._io = BytesIO(self._io).__enter__()
+        return self
+
 
-    def close(self):
-        self._io.close()
+downstreamMapping = {
+    bytes: KaitaiBytesDownStream,
+    str: getFileDownStream,
+    Path: getFileDownStream,
+}
+
+
+def get_downstream(x: typing.Union[bytes, str, Path], *args, **kwargs) -> IKaitaiDownStream:
+    return downstreamMapping[type(x)](x, *args, **kwargs)
```

Comment:
This won't work correctly for multiple reasons:

* On Unix systems, Python accepts both `str` and `bytes` as paths, even in Python 3. This is to allow working with paths that have no consistent encoding. This isn't handled correctly here - the `bytes` path will instead be treated as the contents of a stream.
* The lookup will fail if `type(x)` is a *subclass* of one of the types listed in the mapping. This is unlikely to be a problem for `bytes` and `str`, but is guaranteed to break for `Path`, because there are no instances of `Path` itself (only of its subclasses, `WindowsPath` or `PosixPath`).
* There is no way to create an `IKaitaiDownStream` from an existing IO stream, which means there's no way to implement `KaitaiStream.from_io` based on this function.

Reply:
>On Unix systems, Python accepts both str and bytes as paths, even in Python 3. This is to allow working with paths that have no consistent encoding. This isn't handled correctly here - the bytes path will instead be treated as the contents of a stream.

Thanks. Haven't known about this fact. How are we going to address it? `Path` doesn't seem to support `bytes` at all, so we can consider it as dropped. Should they be encoded into strings somehow?

>The lookup will fail if type(x) is a subclass of one of the types listed in the mapping. This is unlikely to be a problem for bytes and str, but is guaranteed to break for Path, because there are no instances of Path itself (only of its subclasses, WindowsPath or PosixPath).

```python
Path("./a.txt").__class__.mro()

[pathlib.PosixPath,
 pathlib.Path,
 pathlib.PurePosixPath,
 pathlib.PurePath,
 object]
```

* There is no way to create an IKaitaiDownStream from an existing IO stream, which means there's no way to implement KaitaiStream.from_io based on this function.

It's by design. If one needs a stream, he must write a wrapper implementing `IKaitaiDownStream`.